File: CoJam/CJMainViewController.m
//
//  CJMainViewController.m
//  CoJam
//
//  Created by Stephen Greco on 4/23/15.
//  Copyright (c) 2015 Stephen Greco. All rights reserved.
//

#import "CJMainViewController.h"
#import "CJLoginViewController.h"
#import "CJHomeHeaderTableViewCell.h"
#import "CJColors.h"
#import "CJRoomListTableViewCell.h"
#import "CJRoomViewController.h"

#import <Parse/Parse.h>
#import <MBProgressHUD/MBProgressHUD.h>

static NSString* const CJRoomListTableViewCellIdentifier = @"CJRoomListTableViewCell";

@interface CJMainViewController ()

@property CJNewRoomView *createRoomForm;
@property UIView *backgroundTint;

@property PFQuery *roomsQuery;

@end

@implementation CJMainViewController

- (void)viewDidLoad {
    
    [super viewDidLoad];
    [self setNeedsStatusBarAppearanceUpdate];
    [self.navigationController setNavigationBarHidden:NO animated:YES];
    self.view.backgroundColor = [CJColors backgroundColorA];
    
    _tableView = [[UITableView alloc] initWithFrame:self.view.frame];
    _tableView.backgroundColor = [UIColor whiteColor];
    _tableView.delegate = self;
    _tableView.dataSource = self;
    _tableView.rowHeight = 75;
    [self.view addSubview: _tableView];
    
    [_tableView registerClass:[CJRoomListTableViewCell class]
           forCellReuseIdentifier:CJRoomListTableViewCellIdentifier];
    
    CJSpotifyHelper *helper = [CJSpotifyHelper defaultHelper];
    helper.delegate = self;
    
    _roomsQuery = [PFQuery queryWithClassName:@"Room"];
    [_roomsQuery whereKey:@"members" equalTo:[PFUser currentUser]];
    [_roomsQuery whereKey:@"isActive" equalTo:@YES];
    
    [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    
    [helper attemptToReauthenticateWithBlock:^(BOOL success, NSError *error) {
        if (error) {
            NSLog(@"Auth error: %@", [error localizedDescription]);
        }
        
        if (!success) {
            SPTAuthViewController *authvc = [SPTAuthViewController authenticationViewController];
            authvc.delegate = [CJSpotifyHelper defaultHelper];
            authvc.modalPresentationStyle = UIModalPresentationCustom;
            authvc.modalTransitionStyle = UIModalTransitionStyleCrossDissolve;
            self.modalPresentationStyle = UIModalPresentationCurrentContext;
            self.definesPresentationContext = YES;
            [self.navigationController presentViewController:authvc animated:YES completion:nil];
        } else {
            NSLog(@"Logged in as %@", [CJUser currentUser].spotifyUsername);
            [self refresh];
        }
    }];
    
    [self initNavbar];
    
}

- (void) initNavbar {
    // Logo in the center of navigation bar
    self.navigationController.view.window.tintColor = [CJColors altColorB];
    self.navigationController.navigationBar.barTintColor = [CJColors backgroundColorA];
    
    UIImage *logo = [UIImage imageNamed:@"CJLogo"];
    float height = self.navigationController.navigationBar.frame.size.height;
    float imageWidth = logo.size.width;
    float imageHeight = logo.size.height;
    UIView *logoView = [[UIView alloc] initWithFrame:CGRectMake(0, height * 0.15,
                                                                imageWidth * (height * 0.6 / imageHeight),
                                                                height * 0.6)];
    UIImageView *titleImageView = [[UIImageView alloc] initWithImage:logo];
    titleImageView.frame = CGRectMake(0, 0, logoView.frame.size.width, logoView.frame.size.height);
    [logoView addSubview:titleImageView];
    self.navigationItem.titleView = logoView;
    
    UIBarButtonItem *addRoom = [[UIBarButtonItem alloc]
                                initWithBarButtonSystemItem:UIBarButtonSystemItemAdd
                                target:self action:@selector(addRoom)];
    
    UIBarButtonItem *searchRoom = [[UIBarButtonItem alloc]
                                initWithBarButtonSystemItem:UIBarButtonSystemItemSearch
                                target:self action:nil];
    
    self.navigationItem.rightBarButtonItem = addRoom;
    self.navigationItem.leftBarButtonItem = searchRoom;
    
}

- (void) addRoom {
    
    if (_createRoomForm != nil) {
        return;
    }
    
    float width = self.view.frame.size.width;
    float height = self.navigationController.navigationBar.frame.size.height;
    float statusHeight = [UIApplication sharedApplication].statusBarFrame.size.height;
    _createRoomForm = [[CJNewRoomView alloc]
                              initWithFrame:CGRectMake(0,0, width, 1.1 * height)];
    _createRoomForm.delegate = self;
    _backgroundTint = [[UIView alloc] initWithFrame:self.view.frame];
    _backgroundTint.backgroundColor = [UIColor clearColor];
    [self.view addSubview:_backgroundTint];
    
    [self.view addSubview:_createRoomForm];
    [UIView animateWithDuration:.25
                     animations:^{
                         _backgroundTint.backgroundColor = [UIColor colorWithWhite:0.5 alpha:0.5];
                         _createRoomForm.frame = CGRectOffset(_createRoomForm.frame, 0, height + statusHeight);
                     }
                     completion:^(BOOL finished){
                         // WOOOOOO
                     }];
}

- (void) newRoomCreated:(CJRoom *)room {
    [_backgroundTint removeFromSuperview];
    _backgroundTint = nil;
    _createRoomForm = nil;
    [self refresh];
}

- (void) refresh {
    if ([MBProgressHUD HUDForView:self.view] == nil) {
        [MBProgressHUD showHUDAddedTo:self.view animated:YES];
    }
    [_roomsQuery findObjectsInBackgroundWithBlock:^(NSArray *objects, NSError *error) {
        if (error != nil) {
            NSLog(@"%@", [error localizedDescription]);
        }
        NSLog(@"Found %i room(s)", (int)[objects count]);
        _rooms = [[NSArray alloc] initWithArray:objects];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [_tableView reloadData];
            [MBProgressHUD hideHUDForView:self.view animated:YES];
        });
    }];
}

- (void) openRoomView:(CJRoom*) room {
    
    if (room == nil) {
        // TODO: open some sort of 404 view
    } else {
        CJRoomViewController *roomVc = [[CJRoomViewController alloc]
                                        initWithNibName:nil bundle:nil];
        roomVc.room = room;
        [self.navigationController pushViewController:roomVc animated:YES];
    }
    
}


- (NSInteger) numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (CGFloat) tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 75;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return (_rooms == nil) ? 0 : [_rooms count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    CJRoomListTableViewCell *cell = [_tableView dequeueReusableCellWithIdentifier:
                                     CJRoomListTableViewCellIdentifier];
    
    if (cell == nil) {
        cell = [[CJRoomListTableViewCell alloc]
                initWithStyle:UITableViewCellStyleDefault
                reuseIdentifier:CJRoomListTableViewCellIdentifier];
    }
    
    CJRoom *room = _rooms[indexPath.row];
    cell.displayName.text = room.displayName;
    
    return cell;
    
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    
    CJRoom *room = _rooms[indexPath.row];
    [self openRoomView:room];
    
}

-(UIStatusBarStyle) preferredStatusBarStyle{
    return UIStatusBarStyleLightContent;
}

- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    if (_createRoomForm != nil) {
        UIView *responder = resignFirstResponder(self.view);
        if ([responder superview] != _createRoomForm) {
            [_createRoomForm remove];
            [_backgroundTint removeFromSuperview];
            _createRoomForm = nil;
            _backgroundTint = nil;
        }
    }
}

UIView *resignFirstResponder(UIView *theView){
    if([theView isFirstResponder]) {
        [theView resignFirstResponder];
        return theView;
    }
    for(UIView *subview in theView.subviews) {
        UIView *result = resignFirstResponder(subview);
        if(result) return result;
    }
    return nil;
}

@end

File: CoJam/CJMainViewController.swift
//
//  CJMainViewController.swift
//  CoJam
//
//  Created by Stephen Greco on 4/23/15.
//  Copyright (c) 2015 Stephen Greco. All rights reserved.
//

import UIKit

class CJMainViewController: UIViewController, CJSpotifyManagerDelegate,
    CJNewRoomViewDelegate, UITableViewDelegate, UITableViewDataSource,
    UIGestureRecognizerDelegate {

    @IBOutlet weak var newRoomView: UIView!
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var newRoomViewHeightConstraint: NSLayoutConstraint!
    
    var refreshControl : UIRefreshControl!
    var tapRecognizer : UITapGestureRecognizer!
    let helper = CJSpotifyManager.sharedManager
    
    var myRooms : [CJRoom]?
    var myRoomsQuery : PFQuery!
    var nearbyRooms : [CJRoom]?
    var nearbyRoomsQuery : PFQuery!
    
    var showingNewRoomView = false
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        myRoomsQuery = CJRoom.queryWithPredicate(NSPredicate(format: "isActive == true"))
        nearbyRoomsQuery = myRoomsQuery // Temporary
        
        helper.delegate = self
        
        // hide the create room panel
        newRoomViewHeightConstraint.constant = 0
        
        let newRoomVC = childViewControllers[0] as! CJNewRoomViewController
        newRoomVC.delegate = self
        
        
        refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: "refresh", forControlEvents: .ValueChanged)
        tableView.addSubview(refreshControl)
        
        tapRecognizer = UITapGestureRecognizer(target: self, action: "tapped:")
        tapRecognizer.delegate = self
        self.view.addGestureRecognizer(tapRecognizer)
        
        MBProgressHUD.showHUDAddedTo(self.view, animated: true)
        helper.attemptToReauthenticate { (success: Bool, _) -> Void in
            MBProgressHUD.hideHUDForView(self.view, animated: true)
            if success {
                self.refresh()
            } else {
                self.showSpotifyAuthController()
            }
        }
        
    }
    
    override func viewDidAppear(animated: Bool) {
        // Refresh when opened
        NSNotificationCenter.defaultCenter().addObserver(self, selector: "refresh",
            name: UIApplicationDidBecomeActiveNotification, object: nil)
    }
    
    override func viewDidDisappear(animated: Bool) {
        NSNotificationCenter.defaultCenter().removeObserver(self,
            name: UIApplicationDidBecomeActiveNotification, object: nil)
    }
    
    func refresh() {
        self.refreshControl.beginRefreshing()
        myRoomsQuery.findObjectsInBackgroundWithBlock { (objects : [AnyObject]?, _) in
            self.myRooms = objects as? [CJRoom]
            self.nearbyRoomsQuery.findObjectsInBackgroundWithBlock { (objects : [AnyObject]?, _) in
                self.nearbyRooms = objects as? [CJRoom]
                dispatch_async(dispatch_get_main_queue()) {
                    self.tableView.reloadData()
                    if self.refreshControl.refreshing {
                        self.refreshControl.endRefreshing()
                    }
                }
            }
        }
    }
    
    @IBAction func newRoomButtonPressed(sender: AnyObject) {
        self.view.layoutIfNeeded()
        newRoomViewHeightConstraint.constant = 50
        UIView.animateWithDuration(0.2) {
            self.view.layoutIfNeeded()
            self.showingNewRoomView = true
        }
    }
    
    func hideNewRoomView() {
        self.view.layoutIfNeeded()
        NSNotificationCenter.defaultCenter().postNotificationName(
            Notifications.hideKeyboardNotification, object: nil)
        newRoomViewHeightConstraint.constant = 0
        UIView.animateWithDuration(0.2) {
            self.view.layoutIfNeeded()
            self.showingNewRoomView = false
        }
    }
    
    func showSpotifyAuthController() {
        let authvc = SPTAuthViewController.authenticationViewController()
        authvc.delegate = helper
        authvc.modalPresentationStyle = .Custom
        authvc.modalTransitionStyle = .CrossDissolve
        self.modalPresentationStyle = .CurrentContext
        self.definesPresentationContext = true
        self.navigationController!.presentViewController(authvc,
            animated: true, completion: nil)
    }
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "showRoom" {
            let controller = segue.destinationViewController as! CJRoomViewController
            controller.room = (sender as! CJRoomReference).room
        }
        super.prepareForSegue(segue, sender: sender)
    }
    
    // MARK: - UITableViewDelegate
    
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 2
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        switch section {
        case 0: return myRooms?.count ?? 0
        case 1: return nearbyRooms?.count ?? 0
        default: return 0
        }
    }
    
    func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        return 80
    }
    
    func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch section {
        case 0: return "Your rooms"
        case 1: return "Nearby rooms"
        default: return ""
        }
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {

        let cell = tableView.dequeueReusableCellWithIdentifier(CellIdentifiers.roomListCellIdentifier,
            forIndexPath: indexPath) as! CJRoomListTableCell
        
        switch indexPath.section {
        case 0:
            cell.room = myRooms?[indexPath.row]
        case 1:
            cell.room = nearbyRooms?[indexPath.row]
        default:
            break
        }
        
        cell.updateView()
        return cell
    }

    // MARK: - CJSpotifyManagerDelegate
    
    func failedToLogin() {
        // Do nothing for now
    }
    
    func openRoomView(room: CJRoom?) {
        // Do nothing for now
    }
    
    
    // MARK: - UIGestureRecognizerDelegate
    
    func tapped(sender: AnyObject?) {
        hideNewRoomView()
    }
    
    func gestureRecognizer(gestureRecognizer: UIGestureRecognizer, shouldReceiveTouch touch: UITouch) -> Bool {
        let point = touch.locationInView(newRoomView)
        return showingNewRoomView && !CGRectContainsPoint(newRoomView.bounds, point)
    }
    
    // MARK: - CJNewRoomDelegate
    
    func newRoomCreated(room: CJRoom) {
        hideNewRoomView()
        refresh()
    }
    
}

protocol CJRoomReference {
    var room : CJRoom? {get set}
}

class CJRoomListTableCell : UITableViewCell, CJRoomReference {
    
    @IBOutlet weak var albumCover: AsyncImageView!
    @IBOutlet weak var roomName: UILabel!
    @IBOutlet weak var nowPlaying: UILabel!
    var room : CJRoom?
    
    func updateView() {
        roomName.text = room?.displayName ?? "Untitled"
        albumCover.imageURL = NSURL(string: "https://upload.wikimedia.org/wikipedia/en/3/35/The_Eminem_Show.jpg")
    }
    
}
